qiskit/primitives/containers/bindings_array.py

```diff
@@ -218,12 +218,16 @@ def as_array(self, parameters: Iterable[Parameter | str] | None = None) -> np.nd
                 size = arr.shape[-1]
                 ret[..., pos : pos + size] = arr
                 pos += size
-        elif self.kwvals:
+        else:
```

Comment:
Ive been running into a lot of issues testing code with this method, and this change in particular seems to make existing workflows not possible.

The general issues is BindingsArray currently allows mixing and matching between named and unnamed parameter values, but then this array method currently wants only the named parameters to construct an array.

A typical workflow that might want this would be to use a circuits parameters to ensure ordering of this array. Eg
```
param_vals = bindings.as_array(circuit.parameters)
```
But currently this will error unless the bindings is all kwvals. If it has any `vals` it will error.

I can think of two solutions for this issue:

1. We remove vals entirely from BindingsArray or all parameters have to be named. This would probably simplify the implementation, but it changes the types that can be converted to BindingsArrays by its `coerce` method.  In particular it removes the abiility to initialize from an `np.array`. `EstimatorPub` could still coerce though since it also has access to the circuits so could use them in the coercian for a raw array input for param values.
2. We keep vals and kwvals in the init, but don't allow mixing the two. A BindingsArray has to be only defined with 1 or more vals arrays, or only defined with one or more kwvals arrarys, and if you mix them it errors. In this case methods like `as_array` in the case of only vals could only be able to validate that the number of parameters is correct, not the names of the parameters.

Reply:
I'd like to go forward with 1.